# Data Wrangling: American Housing Survey 2015

In this notebook, I read in and examine the survey results from the 2015 American Housing Survey (AHS). The AHS is sponsored by the Department of Housing and Urban Development and conducted by the Census Bureau every odd-numbered year since 1973.

In [1]:
# First, I import pandas and  read in the household.csv file, which contains one entry for each household surveyed. 

import pandas as pd
import numpy as np

df_household = pd.read_csv('Data_Files/household.csv')

In [2]:
# examine the column names and values in the first 10 records of the dataframe

pd.set_option('display.max_columns', None)
df_household.head(10)

CONTROL  TOTROOMS TOTHCAMT  PERPOVLVL JACPRIMARY JACSECNDRY JADEQUACY  \
0  '11000001'         7     '14'        501        '0'        '0'       '2'   
1  '11000002'         7     '15'        203        '0'        '0'       '2'   
2  '11000003'         4     '10'         80        '0'        '0'       '2'   
3  '11000005'         8     '11'        378        '0'        '0'       '2'   
4  '11000006'         5     '05'        266        '2'        '2'       '2'   
5  '11000007'         8     '15'        501        '0'        '0'       '2'   
6  '11000008'         3     '-9'         -9        '2'        '0'       '2'   
7  '11000009'         5     '-9'         -9        '0'        '0'       '2'   
8  '11000010'         5     '08'        185        '0'        '0'       '2'   
9  '11000012'         3     '09'         51        '0'        '0'       '2'   

  JARTACCESS JARTATTRACT JARTAWARE JARTECON JARTIDENT JARTIMPROVE JARTJOBS  \
0        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
1        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
2        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
3        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
4        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
5        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
6        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
7        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
8        '0'         '0'       '0'      '0'       '0'         '0'      '0'   
9        '0'         '0'       '0'      '0'       '0'         '0'      '0'   

  JARTNEARBY JARTSATIS JARTSOC JASTH1MED JASTH2MED JASTHEMR JASTHMA  \
0        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
1        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
2        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
3        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
4        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
5        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
6        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
7        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
8        '0'       '0'     '0'       '0'       '0'      '0'     '0'   
9        '0'       '0'     '0'       '0'       '0'      '0'     '0'   

  JBATHEXCLU JBATHROOMS JBEDROOMS JBLD JCHEMSTOR JCOLD JCOLDCOST JCOLDEQ  \
0        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
1        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
2        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
3        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
4        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
5        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
6        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
7        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
8        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   
9        '0'        '0'       '0'  '0'       '0'   '0'       '0'     '0'   

  JCOLDEQFREQ JCOLDHTCAP JCOLDINSUL JCOLDOTHER JCOLDUTIL JCONDO JCOOKFUEL  \
0         '0'        '0'        '0'        '0'       '0'    '0'       '0'   
1         '0'        '0'        '0'        '0'       '0'    '1'       '0'   
2         '0'        '0'        '0'        '0'       '0'    '0'       '0'   
3         '0'        '0'        '0'        '0'       '0'    '0'       '0'   
4         '0'        '0'        '0'        '0'       '0'    '0'       '2'   
5         '0'        '0'        '0'        '0'       '0'    '0'       '0'   
6         '0'        '0'        '0'        '0'       '0'    '0'       '0'   
7         '0'        '0'        '0'       

The values of this DataFrame are categoricals that have been encoded. They will need to be translated so that the values can be understood when the analysis is performed. In addition, the column names have been abbreviated. It will be easier to perform analysis if the column names are edited as well.

In [3]:
# use df.info() to get more general information on the DataFrame.

df_household.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69493 entries, 0 to 69492
Columns: 1091 entries, CONTROL to REMODAMT
dtypes: float64(483), int64(51), object(557)
memory usage: 578.4+ MB


The DataFrame df_household has almost 70,000 entries with 1,091 attributes for each entry. There are only three data types within this DataFrame. For this particular analysis, only some of the attributes will be examined. The DataFrame for the analysis can be reduced to only the relevant columns using "Column Values.xlsx" which has the original names of the columns that will be used for this analysis, and the new column names as well.

In [4]:
# read in an excel file which contains the names of the columns that will be used in this project, and the new names

cols_to_keep = pd.ExcelFile('Data_Files/Column Values.xlsx')
df_colnames = cols_to_keep.parse('new_colnames')
df_colnames.head(10)

original_col        new_col
0      CONTROL        control
1     TOTROOMS    total_rooms
2     TOTHCAMT   housing_cost
3         RENT   monthly_rent
4       GARAGE         garage
5       DINING     num_dining
6       LAUNDY    num_laundry
7      GUTREHB      gut_rehab
8        CONDO       is_condo
9     STAIRBRK  stairs_broken

In [5]:
# use df_colnames to select the columns from df_household that are going to be used for analysis

df_household_clean = df_household[df_colnames['original_col']]
df_household_clean.head()

CONTROL  TOTROOMS TOTHCAMT  RENT GARAGE  DINING  LAUNDY GUTREHB CONDO  \
0  '11000001'         7     '14'    -9    '1'       1       0     '2'   '2'   
1  '11000002'         7     '15'    -9    '1'       0       0    '-9'   '2'   
2  '11000003'         4     '10'   500    '1'       0       0    '-9'   '2'   
3  '11000005'         8     '11'    -9    '1'       1       2     '2'   '2'   
4  '11000006'         5     '05'    -9    '1'       0       0    '-9'   '2'   

  STAIRBRK STAIRS  MUST  RATINGHS  RATINGNH NHQPCRIME NHQPUBTRN NHQRISK  \
0     '-9'    '2'   '5'        10        10       '2'       '2'     '2'   
1      '2'    '1'   '5'        10        10       '2'       '1'     '2'   
2     '-9'   '-9'  '-9'        -6        -6      '-6'      '-6'    '-6'   
3     '-9'   '-9'  '-9'         9         8       '2'       '2'     '2'   
4     '-9'   '-6'  '-6'        -6        -6      '-6'      '-6'    '-6'   

  NHQSCHOOL NHQSCRIME SUBDIV TENURE PORCH LEAKO LEAKI NOTOIL PAINTPEEL  \
0       '1'       '2'    '1'    '1'   '1'   '2'   '2'    '2'       '2'   
1       '1'       '2'    '2'    '1'   '1'   '2'   '2'    '2'       '2'   
2      '-6'      '-6'   '-6'    '2'   '1'   '2'   '2'    '1'       '2'   
3       '1'       '2'    '1'    '1'   '1'   '1'   '1'    '2'       '2'   
4      '-6'      '-6'   '-6'    '3'   '1'   '2'   '2'    '2'       '2'   

   MONLSTOCC SUITYRRND TIMESHARE RENTCNTRL RENTSUB NEARABAND NEARBARCL  \
0         -9      '-9'      '-9'      '-9'    '-9'       '3'       '3'   
1         -9      '-9'      '-9'      '-9'    '-9'       '3'       '3'   
2         -9      '-9'      '-9'      '-9'     '4'      '-6'      '-6'   
3         -9      '-9'      '-9'      '-9'    '-9'       '3'       '3'   
4         -9      '-9'      '-9'      '-9'     '8'      '-6'      '-6'   

  NEARTRASH INTLANG OMB13CBSA   BLD ENTRYSYS  HHAGE  HHMOVE PARTNER HSHLDTYPE  \
0       '3'    '01'   '37980'  '02'     '-9'     49    2000     '0'       '1'   
1       '3'    '01'   '99998'  '02'     '-9'     77    2005     '0'       '7'   
2      '-6'    '-6'   '99998'  '03'     '-9'     25    2014     '0'       '3'   
3       '3'    '01'   '99998'  '02'     '-9'     69    1995     '0'       '1'   
4      '-6'    '01'   '99998'  '02'     '-9'     21    2015     '0'       '2'   

   NUMPEOPLE  YRBUILT  HOA  STORIES  UNITFLOORS UNITSIZE  BEDROOMS  KITCHENS  \
0          3     2000  '1'        2           2      '6'         3         1   
1          2     2000  '2'        2           2      '8'         3         1   
2          3     1970  '2'        1           1      '3'         2         1   
3          3     1970  '2'        1           1      '6'         4         1   
4          3     1980  '2'        1           1     '-6'         3         1   

  BATHEXCLU KITEXCLU FRIDGE DISHWASH WASHER DRYER FIREPLACE COLD NOWAT  \
0      '-9'     '-9'    '1'      '1'    '1'   '1'       '2'  '2'   '2'   
1      '-9'     '-9'    '1'      '1'    '1'   '1'       '4'  '2'   '2'   
2      '-9'     '-9'    '1'      '1'    '1'   '1'       '4'  '2'   '2'   
3      '-9'     '-9'    '1'      '1'    '1'   '1'       '1'  '2'   '2'   
4      '-9'     '-9'    '1'      '1'    '1'   '1'       '1'  '2'   '2'   

  WALLCRACK FLOORHOLE FNDCRUMB ROOFSHIN ROOFHOLE ROOFSAG WALLSIDE WALLSLOPE  \
0       '2'       '2'      '2'      '2'      '2'     '2'      '2'       '2'   
1       '2'       '2'      '2'      '2'      '2'     '2'      '2'       '2'   
2       '2'       '2'      '2'      '2'      '2'     '2'      '2'       '2'   
3       '2'       '2'      '2'      '2'      '2'     '2'      '2'       '2'   
4       '2'       '2'     '-6'      '2'      '2'     '2'      '2'       '2'   

  WINBOARD WINBROKE WINBARS RODENT ROACH SEWBREAK FUSEBLOW HRATE NRATE  \
0      '2'      '2'     '2'    '5'   '5'      '6'      '6'  '-9'  '-9'   
1      '2'      '2'     '2'    '5'   '5'      '6'      '6'  '-9'  '-9'   
2      '2'      '2'     '2'    '5'   '5'      '6'      '6'  '-6'  '-6'   
3     

In [6]:
# examine the shape of the new DataFrame, df_household_clean. The DataFrame has only 83 columns now.

df_household_clean.shape

(69493, 83)

In [7]:
# use df.info() to get more general information on the DataFrame. All of the values are stored as int or python objects.

df_household_clean.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69493 entries, 0 to 69492
Data columns (total 83 columns):
CONTROL       69493 non-null object
TOTROOMS      69493 non-null int64
TOTHCAMT      69493 non-null object
RENT          69493 non-null int64
GARAGE        69493 non-null object
DINING        69493 non-null int64
LAUNDY        69493 non-null int64
GUTREHB       69493 non-null object
CONDO         69493 non-null object
STAIRBRK      69493 non-null object
STAIRS        69493 non-null object
MUST          69493 non-null object
RATINGHS      69493 non-null int64
RATINGNH      69493 non-null int64
NHQPCRIME     69493 non-null object
NHQPUBTRN     69493 non-null object
NHQRISK       69493 non-null object
NHQSCHOOL     69493 non-null object
NHQSCRIME     69493 non-null object
SUBDIV        69493 non-null object
TENURE        69493 non-null object
PORCH         69493 non-null object
LEAKO         69493 non-null object
LEAKI         69493 non-null object
NOTOIL        69493 non-null obje

In [8]:
# use .strip() to remove all of the single quotation marks surrounding all of the values.

for column in df_household_clean:
    if df_household_clean[column].values.dtype == 'O':
        df_household_clean[column] = df_household_clean[column].map(lambda x: x.strip("'"))

C:\Users\wongk3.NA\AppData\Local\Continuum\Anaconda2\envs\newPy36\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [9]:
# use .head() to examine the first 5 rows of the DataFrame and confirm no more single quotes remain

df_household_clean.head()

CONTROL  TOTROOMS TOTHCAMT  RENT GARAGE  DINING  LAUNDY GUTREHB CONDO  \
0  11000001         7       14    -9      1       1       0       2     2   
1  11000002         7       15    -9      1       0       0      -9     2   
2  11000003         4       10   500      1       0       0      -9     2   
3  11000005         8       11    -9      1       1       2       2     2   
4  11000006         5       05    -9      1       0       0      -9     2   

  STAIRBRK STAIRS MUST  RATINGHS  RATINGNH NHQPCRIME NHQPUBTRN NHQRISK  \
0       -9      2    5        10        10         2         2       2   
1        2      1    5        10        10         2         1       2   
2       -9     -9   -9        -6        -6        -6        -6      -6   
3       -9     -9   -9         9         8         2         2       2   
4       -9     -6   -6        -6        -6        -6        -6      -6   

  NHQSCHOOL NHQSCRIME SUBDIV TENURE PORCH LEAKO LEAKI NOTOIL PAINTPEEL  \
0         1         2      1      1     1     2     2      2         2   
1         1         2      2      1     1     2     2      2         2   
2        -6        -6     -6      2     1     2     2      1         2   
3         1         2      1      1     1     1     1      2         2   
4        -6        -6     -6      3     1     2     2      2         2   

   MONLSTOCC SUITYRRND TIMESHARE RENTCNTRL RENTSUB NEARABAND NEARBARCL  \
0         -9        -9        -9        -9      -9         3         3   
1         -9        -9        -9        -9      -9         3         3   
2         -9        -9        -9        -9       4        -6        -6   
3         -9        -9        -9        -9      -9         3         3   
4         -9        -9        -9        -9       8        -6        -6   

  NEARTRASH INTLANG OMB13CBSA BLD ENTRYSYS  HHAGE  HHMOVE PARTNER HSHLDTYPE  \
0         3      01     37980  02       -9     49    2000       0         1   
1         3      01     99998  02       -9     77    2005       0         7   
2        -6      -6     99998  03       -9     25    2014       0         3   
3         3      01     99998  02       -9     69    1995       0         1   
4        -6      01     99998  02       -9     21    2015       0         2   

   NUMPEOPLE  YRBUILT HOA  STORIES  UNITFLOORS UNITSIZE  BEDROOMS  KITCHENS  \
0          3     2000   1        2           2        6         3         1   
1          2     2000   2        2           2        8         3         1   
2          3     1970   2        1           1        3         2         1   
3          3     1970   2        1           1        6         4         1   
4          3     1980   2        1           1       -6         3         1   

  BATHEXCLU KITEXCLU FRIDGE DISHWASH WASHER DRYER FIREPLACE COLD NOWAT  \
0        -9       -9      1        1      1     1         2    2     2   
1        -9       -9      1        1      1     1         4    2     2   
2        -9       -9      1        1      1     1         4    2     2   
3        -9       -9      1        1      1     1         1    2     2   
4        -9       -9      1        1      1     1         1    2     2   

  WALLCRACK FLOORHOLE FNDCRUMB ROOFSHIN ROOFHOLE ROOFSAG WALLSIDE WALLSLOPE  \
0         2         2        2        2        2       2        2         2   
1         2         2        2        2        2       2        2         2   
2         2         2        2        2        2       2        2         2   
3         2         2        2        2        2       2        2         2   
4         2         2       -6        2        2       2        2         2   

  WINBOARD WINBROKE WINBARS RODENT ROACH SEWBREAK FUSEBLOW HRATE NRATE  \
0        2        2       2      5     5        6        6    -9    -9   
1        2        2       2      5     5        6        6    -9    -9   
2        2        2       2      5     5        6        6    -6    -6   
3        2        2       2      4     5

In [10]:
# read in "AHS 2015 Value Labels.csv" which has all of the labels for the values in the DataFrame

df_valuelabels = pd.read_csv('Data_Files/AHS 2015 Value Labels.csv', encoding='ISO-8859-1')
df_valuelabels = df_valuelabels.set_index('NAME')
df_valuelabels = df_valuelabels[df_valuelabels.Table == 'HOUSEHOLD']
df_valuelabels.head()

Table    FMTName Type Value  \
NAME                                         
ACPRIMARY  HOUSEHOLD  ACPRIMARY    C    01   
ACPRIMARY  HOUSEHOLD  ACPRIMARY    C    02   
ACPRIMARY  HOUSEHOLD  ACPRIMARY    C    03   
ACPRIMARY  HOUSEHOLD  ACPRIMARY    C    04   
ACPRIMARY  HOUSEHOLD  ACPRIMARY    C    05   

                                                       Label FLAT METRO  
NAME                                                                     
ACPRIMARY   Electric powered central air conditioning system   NO   YES  
ACPRIMARY  Piped gas powered central air conditioning system   NO   YES  
ACPRIMARY     LP gas powered central air conditioning system   NO   YES  
ACPRIMARY  Other fuel source powered air conditioning system   NO   YES  
ACPRIMARY                             1 room air conditioner   NO   YES

In [11]:
# set the index of df_household_clean to the "CONTROL" column

df_household_clean = df_household_clean.set_index('CONTROL')

In [12]:
# loop over df_household_clean's columns and map the values from df_valuelabels into the columns using the .map() method

for column in df_household_clean:
    if df_household_clean[column].values.dtype != 'int64' and df_household_clean[column].name != 'OMB13CBSA':
        dict_values = df_valuelabels.loc[column].set_index('Value')['Label'].to_dict()
        df_household_clean[column] = df_household_clean[column].map(dict_values)
        

In [13]:
# examine the first five records of df_household_clean to confirm the values have been mapped correctly

df_household_clean.head()

TOTROOMS      TOTHCAMT  RENT GARAGE  DINING  LAUNDY GUTREHB CONDO  \
CONTROL                                                                       
11000001         7  1250 to 1499    -9    Yes       1       0      No    No   
11000002         7  1500 to 1999    -9    Yes       0       0     NaN    No   
11000003         4    600 to 699   500    Yes       0       0     NaN    No   
11000005         8    700 to 799    -9    Yes       1       2      No    No   
11000006         5    300 to 349    -9    Yes       0       0     NaN    No   

         STAIRBRK STAIRS   MUST  RATINGHS  RATINGNH NHQPCRIME NHQPUBTRN  \
CONTROL                                                                   
11000001      NaN     No  Never        10        10  Disagree  Disagree   
11000002       No    Yes  Never        10        10  Disagree     Agree   
11000003      NaN    NaN    NaN        -6        -6       NaN       NaN   
11000005      NaN    NaN    NaN         9         8  Disagree  Disagree   
11000006      NaN    NaN    NaN        -6        -6       NaN       NaN   

           NHQRISK NHQSCHOOL NHQSCRIME SUBDIV  \
CONTROL                                         
11000001  Disagree     Agree  Disagree    Yes   
11000002  Disagree     Agree  Disagree     No   
11000003       NaN       NaN       NaN    NaN   
11000005  Disagree     Agree  Disagree    Yes   
11000006       NaN       NaN       NaN    NaN   

                                                     TENURE PORCH LEAKO LEAKI  \
CONTROL                                                                         
11000001  Owned or being bought by someone in your house...   Yes    No    No   
11000002  Owned or being bought by someone in your house...   Yes    No    No   
11000003                                             Rented   Yes    No    No   
11000005  Owned or being bought by someone in your house...   Yes   Yes   Yes   
11000006                   Occupied without payment of rent   Yes    No    No   

         NOTOIL PAINTPEEL  MONLSTOCC SUITYRRND TIMESHARE RENTCNTRL  \
CONTROL                                                              
11000001     No        No         -9       NaN       NaN       NaN   
11000002     No        No         -9       NaN       NaN       NaN   
11000003    Yes        No         -9       NaN       NaN       NaN   
11000005     No        No         -9       NaN       NaN       NaN   
11000006     No        No         -9       NaN       NaN       NaN   

                                 RENTSUB                          NEARABAND  \
CONTROL                                                                       
11000001                             NaN  There are no vandalized buildings   
11000002                             NaN  There are no vandalized buildings   
11000003        Other government subsidy                                NaN   
11000005                             NaN  There are no vandalized buildings   
11000006  No rental subsidy or reduction                                NaN   

                       NEARBARCL NEARTRASH  INTLANG OMB13CBSA  \
CONTROL                                                         
11000001  No buildings with bars  No trash  English     37980   
11000002  No buildings with bars  No trash  English     99998   
11000003                     NaN       NaN      NaN     99998   
11000005  No buildings with bars  No trash  English     99998   
11000006                     NaN       NaN  English     99998   

                                 BLD ENTRYSYS  HHAGE  HHMOVE  \
CONTROL                                                        
11000001  One-family house, detached      NaN     49    2000   
11000002  One-family house, detached      NaN     77    2005   
11000003  One-family house, attached      NaN     25    2014   
11000005  One-family house, detached      NaN     69    1995   
11000006  One-family house, detached      NaN     21    2015   

                                    PARTNER  \
CONTROL              

In [14]:
# the column OMB13CBSA did not have values in the Value Labels csv, so a dictionary is created to map the metropolitan regions

dict_metro_codes = {'35620':"NY", 
                    '31080':"LA", 
                    '16980':"Chicago",
                    '19100':"Dallas",
                    '37980':"Philadelphia",
                    '26420':"Houston",
                    '47900':"DC",
                    '33100':"Miami",
                    '12060':"Atlanta",
                    '14460':"Boston",
                    '41860':"San Francisco",
                    '19820':"Detroit",
                    '40140':"Riverside",
                    '38060':"Phoenix",
                    '42660':"Seattle",
                    '19740':"Denver",
                    '38300':"Pittsburgh",
                    '38900':"Portland",
                    '17140':"Cincinnati",
                    '17460':"Cleveland",
                    '28140':"Kansas City",
                    '33340':"Milwaukee",
                    '32820':"Memphis",
                    '35380':"New Orleans",
                    '39580':"Raleigh",
                    '99998':"Other Metropolitan",
                    '99999':"Non-Metropolitan"}

In [15]:
# use dict_metro_codes to map the correct metropolitan region name to the column OMB13CBSA

df_household_clean['OMB13CBSA'] = df_household_clean['OMB13CBSA'].map(dict_metro_codes)

In [16]:
# use the new column names in df_colnames to rename all of the columns in df_household_clean

df_household_clean.columns = df_colnames['new_col'][1:]
df_household_clean.head()

new_col   total_rooms  housing_cost  monthly_rent garage  num_dining  \
CONTROL                                                                
11000001            7  1250 to 1499            -9    Yes           1   
11000002            7  1500 to 1999            -9    Yes           0   
11000003            4    600 to 699           500    Yes           0   
11000005            8    700 to 799            -9    Yes           1   
11000006            5    300 to 349            -9    Yes           0   

new_col   num_laundry gut_rehab is_condo stairs_broken stairs  musty  \
CONTROL                                                                
11000001            0        No       No           NaN     No  Never   
11000002            0       NaN       No            No    Yes  Never   
11000003            0       NaN       No           NaN    NaN    NaN   
11000005            2        No       No           NaN    NaN    NaN   
11000006            0       NaN       No           NaN    NaN    NaN   

new_col   rating_house  rating_neighborhood petty_crime near_transit  \
CONTROL                                                                
11000001            10                   10    Disagree     Disagree   
11000002            10                   10    Disagree        Agree   
11000003            -6                   -6         NaN          NaN   
11000005             9                    8    Disagree     Disagree   
11000006            -6                   -6         NaN          NaN   

new_col  risk_of_flood good_schools serious_crime subdivision  \
CONTROL                                                         
11000001      Disagree        Agree      Disagree         Yes   
11000002      Disagree        Agree      Disagree          No   
11000003           NaN          NaN           NaN         NaN   
11000005      Disagree        Agree      Disagree         Yes   
11000006           NaN          NaN           NaN         NaN   

new_col                                              tenure porch  \
CONTROL                                                             
11000001  Owned or being bought by someone in your house...   Yes   
11000002  Owned or being bought by someone in your house...   Yes   
11000003                                             Rented   Yes   
11000005  Owned or being bought by someone in your house...   Yes   
11000006                   Occupied without payment of rent   Yes   

new_col  out_water_leaks in_water_leaks toilet_broke paint_peeling  \
CONTROL                                                              
11000001              No             No           No            No   
11000002              No             No           No            No   
11000003              No             No          Yes            No   
11000005             Yes            Yes           No            No   
11000006              No             No           No            No   

new_col   months_permanent suitable_yrround timeshare rent_control  \
CONTROL                                                              
11000001                -9              NaN       NaN          NaN   
11000002                -9              NaN       NaN          NaN   
11000003                -9              NaN       NaN          NaN   
11000005                -9              NaN       NaN          NaN   
11000006                -9              NaN       NaN          NaN   

new_col                     rent_subsidy                     near_abandoned  \
CONTROL                                                                       
11000001                             NaN  There are no vandalized buildings   
11000002                             NaN  There are no vandalized buildings   
11000003        Other government subsidy                                NaN   
11000005                             NaN  There are no vandalized buildings   
11000006  No rental subsidy or reduction                                NaN   

new_col  

In [17]:
# replace the -9 and -6 values in df_household_clean with NaNs

df_household_clean.replace(-9, np.NaN, inplace=True)
df_household_clean.replace(-6, np.NaN, inplace=True)

In [18]:
# examine the number of null values in every column

df_household_clean.isnull().sum()


new_col
total_rooms                  0
housing_cost              9006
monthly_rent             42102
garage                     327
num_dining                   0
num_laundry                  0
gut_rehab                38063
is_condo                     0
stairs_broken            55571
stairs                   40111
musty                    40283
rating_house             11116
rating_neighborhood      11219
petty_crime               4204
near_transit              5304
risk_of_flood             3284
good_schools              9078
serious_crime             3812
subdivision              24795
tenure                    9006
porch                        0
out_water_leaks           9580
in_water_leaks            9583
toilet_broke              9618
paint_peeling                0
months_permanent         61969
suitable_yrround         68120
timeshare                61189
rent_control             65607
rent_subsidy             41795
                         ...  
dishwash                     0


In [19]:
# use dropna() to drop any records that have null values in the 'rating_house' and 'rating_neighborhood' columns

df_household_clean = df_household_clean.dropna(subset = ['rating_house', 'rating_neighborhood'])

In [20]:
# use dropna() to drop columns with all null values

df_household_clean = df_household_clean.dropna(axis=1, how='all')

In [21]:
df_household_clean.isnull().sum()

new_col
total_rooms                  0
housing_cost                 0
monthly_rent             34868
garage                      36
num_dining                   0
num_laundry                  0
gut_rehab                27849
is_condo                     0
stairs_broken            44410
stairs                   29253
musty                    29276
rating_house                 0
rating_neighborhood          0
petty_crime               1007
near_transit              2165
risk_of_flood              579
good_schools              5062
serious_crime              690
subdivision              18930
tenure                       0
porch                        0
out_water_leaks             79
in_water_leaks              75
toilet_broke               105
paint_peeling                0
rent_control             54727
rent_subsidy             34270
near_abandoned             291
near_bar_windows           793
near_trash                  84
                         ...  
dishwash                     0


In [40]:
df_household_clean = df_household_clean.drop(['household_income', 'family_income'], axis=1)

In [41]:
# the resulting DataFrame has 58,233 entries, 77 columns, and an index column

df_household_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 58233 entries, 11000001 to 11085287
Data columns (total 77 columns):
total_rooms              58233 non-null int64
housing_cost             58233 non-null object
monthly_rent             23365 non-null float64
garage                   58197 non-null object
num_dining               58233 non-null int64
num_laundry              58233 non-null int64
gut_rehab                30384 non-null object
is_condo                 58233 non-null object
stairs_broken            13823 non-null object
stairs                   28980 non-null object
musty                    28957 non-null object
rating_house             58233 non-null float64
rating_neighborhood      58233 non-null float64
petty_crime              57226 non-null object
near_transit             56068 non-null object
risk_of_flood            57654 non-null object
good_schools             53171 non-null object
serious_crime            57543 non-null object
subdivision              39303 non-null 

In [42]:
# save the final dataframe to csv

df_household_clean.to_csv('Data_Files/df_household_clean.csv')